product/src/main/java/com/yas/product/service/ProductService.java

```diff
@@ -475,20 +498,53 @@ private void updateProductRelations(ProductPutVm productPutVm, Product product)
-                                                            Map<Long, ProductOption> optionsById) {
-        List<ProductOptionValue> productOptionValues = new ArrayList<>();
-        productPutVm.productOptionValues().forEach(optionValueVm -> optionValueVm.value().forEach(value -> {
-            ProductOptionValue optionValue = ProductOptionValue.builder()
-                .product(product)
-                .displayOrder(optionValueVm.displayOrder())
-                .displayType(optionValueVm.displayType())
-                .productOption(optionsById.get(optionValueVm.productOptionId()))
-                .value(value)
-                .build();
-            productOptionValues.add(optionValue);
-        }));
-        productOptionValueRepository.saveAll(productOptionValues);
-        return productOptionValues;
+                                                               Map<Long, ProductOption> optionsById) {
+        List<ProductOptionValuePutVm> productOptionValues = productPutVm.productOptionValues();
+        List<ProductOptionValueProjection> savedOptionValues
+            = productOptionValueRepository.findAllProjectedByProduct(product);
+        Map<Long, Set<String>> optionValuesMap = new HashMap<>();
+        for (ProductOptionValueProjection optionValue : savedOptionValues) {
+            Long productOptionId = optionValue.getProductOption().getId();
+            Set<String> set;
+            if (optionValuesMap.containsKey(productOptionId)) {
+                set = optionValuesMap.get(productOptionId);
+            } else {
+                set = new HashSet<>();
+                optionValuesMap.put(productOptionId, set);
+            }
+            set.add(optionValue.getValue());
+        }
+
+        List<ProductOptionValue> optionValues = new ArrayList<>();
+
+        for (ProductOptionValuePutVm productOptionValuePutVm : productOptionValues) {
+            Long productOptionId = productOptionValuePutVm.productOptionId();
+            List<String> values = productOptionValuePutVm.value();
+
+            Set<String> set;
+            if (optionValuesMap.containsKey(productOptionId)) {
+                set = optionValuesMap.get(productOptionId);
+            } else {
+                set = new HashSet<>();
+                optionValuesMap.put(productOptionId, set);
+            }
+
+            for (String value : values) {
+                if (!set.contains(value)) {
+                    ProductOptionValue optionValue = ProductOptionValue.builder()
+                        .product(product)
+                        .displayOrder(productOptionValuePutVm.displayOrder())
+                        .displayType(productOptionValuePutVm.displayType())
+                        .productOption(optionsById.get(productOptionId))
+                        .value(value)
+                        .build();
+                    optionValues.add(optionValue);
+                    set.add(value);
+                }
+            }
+        }
```

Comment:
I see that there is a similar code in the create and update method. Can you move it to a private method to reuse?

```
for (ProductOptionValuePostVm productOptionValuePostVm : productOptionValues) {
            Long productOptionId = productOptionValuePostVm.productOptionId();
            List<String> values = productOptionValuePostVm.value();

            Set<String> set;
            if (optionValuesMap.containsKey(productOptionId)) {
                set = optionValuesMap.get(productOptionId);
            } else {
                set = new HashSet<>();
                optionValuesMap.put(productOptionId, set);
            }

            for (String value : values) {
                if (!set.contains(value)) {
                    ProductOptionValue optionValue = ProductOptionValue.builder()
                        .product(savedMainProduct)
                        .displayOrder(productOptionValuePostVm.displayOrder())
                        .displayType(productOptionValuePostVm.displayType())
                        .productOption(optionsById.get(productOptionId))
                        .value(value)
                        .build();
                    optionValues.add(optionValue);
                    set.add(value);
                }
            }
        }
```

Reply:
I have made a method for the similar code. But this would combine 2 post and put vm into one, but i dont think it would make a different for future development, cause those 2 vm are similar anyway